src/Components/WebAssembly/Server/src/DebugProxyLauncher.cs

```diff
@@ -0,0 +1,119 @@
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.Extensions.CommandLineUtils;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.Builder
+{
+    internal static class DebugProxyLauncher
+    {
+        private static readonly object LaunchLock = new object();
+        private static readonly TimeSpan DebugProxyLaunchTimeout = TimeSpan.FromSeconds(10);
+        private static Task<string> LaunchedDebugProxyUrl;
+
+        public static Task<string> EnsureLaunchedAndGetUrl(IServiceProvider serviceProvider)
+        {
+            lock (LaunchLock)
+            {
+                if (LaunchedDebugProxyUrl == null)
+                {
+                    LaunchedDebugProxyUrl = LaunchAndGetUrl(serviceProvider);
+                }
+
+                return LaunchedDebugProxyUrl;
+            }
+        }
+
+        private static async Task<string> LaunchAndGetUrl(IServiceProvider serviceProvider)
+        {
+            var tcs = new TaskCompletionSource<string>();
+
+            var environment = serviceProvider.GetRequiredService<IWebHostEnvironment>();
+            var executablePath = LocateDebugProxyExecutable(environment);
+            var muxerPath = DotNetMuxer.MuxerPathOrDefault();
+            var ownerPid = Process.GetCurrentProcess().Id;
+            var processStartInfo = new ProcessStartInfo
+            {
+                FileName = muxerPath,
+                Arguments = $"exec \"{executablePath}\" --owner-pid {ownerPid}",
+                UseShellExecute = false,
+                RedirectStandardOutput = true,
+            };
+            RemoveUnwantedEnvironmentVariables(processStartInfo.Environment);
+
+            var debugProxyProcess = Process.Start(processStartInfo);
+            var nowListeningRegex = new Regex("^\\s*Now listening on: (.*)$");
+            debugProxyProcess.OutputDataReceived += (sender, eventArgs) =>
+            {
+                var match = nowListeningRegex.Match(eventArgs.Data);
+                if (match.Success)
+                {
+                    var url = match.Groups[1].Captures[0].Value;
+                    tcs.TrySetResult(url);
+                }
+            };
+            debugProxyProcess.BeginOutputReadLine();
+
+            _ = Task.Run(async () =>
+            {
+                await Task.Delay(DebugProxyLaunchTimeout);
+                tcs.TrySetException(new TimeoutException($"Failed to start the debug proxy within the timeout period of {DebugProxyLaunchTimeout.TotalSeconds} seconds."));
+            });
+
+            return await tcs.Task;
+        }
+
+        private static void RemoveUnwantedEnvironmentVariables(IDictionary<string, string> environment)
```

Comment:
Can we have @Tratcher review this?

Reply:
Yes but unless that happens particularly quickly, let's not block the initial merge on it.

It does make sense not to pass through the `ASPNETCore_*` env vars (since we don't want the child process to try using the same ports as the parent, for example).

Reply:
Do we have the list of environment variables before filtering? That would help us understand which ones we should remove and which ones we should preserve, but I believe they all should be removed.

Reply:
You fully control the process being started, correct? In that case it would be cleaner to control the config from within that app. E.g. Don't use `Host.CreateDefaultBuilder` in that app and never add environment variables as a config source.

Reply:
Thanks @Tratcher. We'll track your suggestions as follow up items